// appium-remote-debugger/build/test/unit/remote-messages-specs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const remote_messages_1 = require("../../lib/rpc/remote-messages");
const helpers_1 = require("../helpers/helpers");
const chai_1 = require("chai");
describe('RemoteMessages', function () {
    this.timeout(helpers_1.MOCHA_TIMEOUT);
    const remoteMessages = new remote_messages_1.RemoteMessages();
    describe('getRemoteCommand', function () {
        const commands = [
            'setConnectionKey', 'connectToApp', 'setSenderKey', 'indicateWebView',
            'Runtime.evaluate', 'Runtime.callFunctionOn', 'Page.navigate', 'Page.enable', 'Timeline.start',
            'Timeline.stop',
        ];
        for (const command of commands) {
            it(`should be able to retrieve ${command} command`, function () {
                const remoteCommand = remoteMessages.getRemoteCommand(command, {
                    id: 'test-id',
                    connId: 'test-conn-id',
                    appIdKey: 'test-app-id',
                    pageIdKey: 'test-page-id',
                    senderId: 'test-sender-id',
                    bundleId: 'test.bundle.id',
                });
                (0, chai_1.expect)(remoteCommand).to.be.an.instanceof(Object);
                (0, chai_1.expect)(remoteCommand.__argument).to.exist;
                (0, chai_1.expect)(remoteCommand.__selector).to.exist;
            });
        }
    });
});
