// appium-remote-debugger/build/test/unit/utils-specs.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../../lib/utils");
const lodash_1 = __importDefault(require("lodash"));
const helpers_1 = require("../helpers/helpers");
const chai_1 = require("chai");
describe('utils', function () {
    this.timeout(helpers_1.MOCHA_TIMEOUT);
    describe('appInfoFromDict', function () {
        it('should return the id and entry for a dict', function () {
            const dict = {
                WIRApplicationIdentifierKey: '42',
                WIRApplicationNameKey: 'App Name',
                WIRApplicationBundleIdentifierKey: 'app.name',
                WIRIsApplicationProxyKey: 'false',
                WIRHostApplicationIdentifierKey: '43'
            };
            const [id, entry] = (0, utils_1.appInfoFromDict)(dict);
            (0, chai_1.expect)(id).to.equal(dict.WIRApplicationIdentifierKey);
            (0, chai_1.expect)(entry.id).to.equal(dict.WIRApplicationIdentifierKey);
            (0, chai_1.expect)(entry.name).to.equal(dict.WIRApplicationNameKey);
            (0, chai_1.expect)(entry.bundleId).to.equal(dict.WIRApplicationBundleIdentifierKey);
            (0, chai_1.expect)(entry.isProxy).to.equal(dict.WIRIsApplicationProxyKey === 'true');
            (0, chai_1.expect)(entry.hostId).to.equal(dict.WIRHostApplicationIdentifierKey);
        });
    });
    describe('pageArrayFromDict', function () {
        const basePageDict = {
            1: {
                WIRTitleKey: 'Appium/welcome',
                WIRTypeKey: 'WIRTypeWeb',
                WIRURLKey: 'http://127.0.0.1:4723/welcome',
                WIRPageIdentifierKey: 1,
                WIRConnectionIdentifierKey: 'e777f792-c41e-4e5d-8722-68393af663b2'
            }
        };
        it('should return a valid page array', function () {
            const pageArray = (0, utils_1.pageArrayFromDict)(basePageDict);
            (0, chai_1.expect)(pageArray).to.have.length(1);
        });
        it('should return a valid page array with 13.4-style type key', function () {
            const pageDict = lodash_1.default.defaults({
                2: {
                    WIRTypeKey: 'WIRTypeWebPage'
                }
            }, basePageDict);
            const pageArray = (0, utils_1.pageArrayFromDict)(pageDict);
            (0, chai_1.expect)(pageArray).to.have.length(2);
        });
        it('should not count WIRTypeWeb entries', function () {
            const pageDict = lodash_1.default.defaults({
                2: {
                    WIRTypeKey: 'WIRTypeJavaScript'
                }
            }, basePageDict);
            const pageArray = (0, utils_1.pageArrayFromDict)(pageDict);
            (0, chai_1.expect)(pageArray).to.have.length(1);
        });
    });
    describe('checkParams', function () {
        it('should not throw error when not missing parameters', function () {
            (0, utils_1.checkParams)({ one: 'first', two: 'second', three: 'third' });
        });
        it('should throw error when parameter is missing', function () {
            (0, chai_1.expect)(() => (0, utils_1.checkParams)({ one: 'first', two: null, three: 'third' })).to.throw('Missing parameter: two');
        });
    });
});
