// @moroo/wdio-slack-reporter/dist/constants.js
const SLACK_NAME = "WebdriverIO Reporter";
const SLACK_ICON_URL = "https://webdriver.io/img/webdriverio.png";
const SUCCESS_COLOR = "#36a64f";
const FAILED_COLOR = "#dc3545";
const DEFAULT_COLOR = "#D3D3D3";
const FINISHED_COLOR = "#4366c7";
const DEFAULT_INDENT = "	";
const EMOJI_SYMBOLS = {
  PASSED: ":white_check_mark:",
  SKIPPED: ":double_vertical_bar:",
  PENDING: ":grey_question:",
  FAILED: ":x:",
  ROCKET: ":rocket:",
  CHECKERED_FLAG: ":checkered_flag:",
  STOPWATCH: ":stopwatch:"
};
const SLACK_REQUEST_TYPE = {
  WEB_API_POST_MESSAGE: "web-api:message",
  WEB_API_UPLOAD: "web-api:upload",
  WEBHOOK_SEND: "webhook:send"
};
const EVENTS = {
  POST_MESSAGE: "slackReporter:postMessage",
  UPLOAD: "slackReporter:upload",
  SEND: "slackReporter:send",
  RESULT: "slackReporter:result",
  SCREENSHOT: "slackReporter:screenshot"
};
const ERROR_MESSAGES = {
  UNDEFINED_SLACK_OPTION: "Slack Option is undefined. Please Check Slack Option.",
  NOT_USING_WEBHOOK: "Not using webhook.",
  NOT_USING_WEB_API: "Not using web-api.",
  DISABLED_OPTIONS: "Disabled notifyFailedCase or uploadScreenshotOfFailedCase options."
};
export {
  DEFAULT_COLOR,
  DEFAULT_INDENT,
  EMOJI_SYMBOLS,
  ERROR_MESSAGES,
  EVENTS,
  FAILED_COLOR,
  FINISHED_COLOR,
  SLACK_ICON_URL,
  SLACK_NAME,
  SLACK_REQUEST_TYPE,
  SUCCESS_COLOR
};
