// appium-remote-debugger/build/test/unit/mixins/connect-specs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const connect_1 = require("../../../lib/mixins/connect");
const helpers_1 = require("../../helpers/helpers");
const remote_debugger_1 = require("../../../lib/remote-debugger");
const chai_1 = require("chai");
describe('connect', function () {
    this.timeout(helpers_1.MOCHA_TIMEOUT);
    let rd;
    this.beforeEach(function () {
        rd = new remote_debugger_1.RemoteDebugger();
    });
    describe('getPossibleDebuggerAppKeys', function () {
        it('should return the app key of the specified bundleIds', function () {
            rd._appDict = {
                ['42']: {
                    id: '42',
                    bundleId: 'io.appium.bundle1',
                    isProxy: false,
                    name: 'Bundle1',
                    isActive: true,
                    isAutomationEnabled: true
                },
                ['43']: {
                    id: '43',
                    bundleId: 'io.appium.bundle2',
                    isProxy: false,
                    name: 'Bundle2',
                    isActive: true,
                    isAutomationEnabled: true
                },
                ['44']: {
                    id: '44',
                    bundleId: 'io.appium.bundle3',
                    isProxy: false,
                    name: 'Bundle3',
                    isActive: true,
                    isAutomationEnabled: true
                },
            };
            (0, chai_1.expect)(connect_1.getPossibleDebuggerAppKeys.bind(rd)(['io.appium.bundle1', 'io.appium.bundle2']))
                .to.eql(['42', '43']);
        });
        const webviewBundleIds = [
            'com.apple.WebKit.WebContent',
            'process-com.apple.WebKit.WebContent',
            'process-SafariViewService',
            'com.apple.SafariViewService',
        ];
        for (const webviewBundleId of webviewBundleIds) {
            it(`should return the app key of ${webviewBundleId}`, function () {
                rd._appDict = {
                    ['42']: {
                        id: '42',
                        bundleId: webviewBundleId,
                        isProxy: false,
                        name: 'WebView',
                        isActive: true,
                        isAutomationEnabled: true
                    }
                };
                (0, chai_1.expect)(connect_1.getPossibleDebuggerAppKeys.bind(rd)([])).to.eql(['42']);
            });
        }
        it('should return the app key for the bundleIds when proxied', function () {
            rd._appDict = {
                ['42']: {
                    id: '42',
                    bundleId: 'io.appium.bundle',
                    isProxy: false,
                    name: 'Bundle',
                    isActive: true,
                    isAutomationEnabled: true
                },
                ['43']: {
                    id: '43',
                    bundleId: 'io.appium.proxied.bundle',
                    isProxy: true,
                    hostId: '42',
                    name: 'ProxiedBundle',
                    isActive: true,
                    isAutomationEnabled: true
                }
            };
            (0, chai_1.expect)(connect_1.getPossibleDebuggerAppKeys.bind(rd)(['io.appium.bundle'])).to.eql(['42', '43']);
        });
        it('should return an empty array when there is no appropriate app', function () {
            rd._appDict = {};
            (0, chai_1.expect)(connect_1.getPossibleDebuggerAppKeys.bind(rd)('io.appium.bundle')).to.eql([]);
        });
        it('should return the all app keys when the bundlIds array includes a wildcard', function () {
            rd._appDict = {
                ['42']: {
                    id: '42',
                    bundleId: 'io.appium.bundle1',
                    isProxy: false,
                    name: 'Bundle1',
                    isActive: true,
                    isAutomationEnabled: true
                },
                ['43']: {
                    id: '43',
                    bundleId: 'io.appium.bundle2',
                    isProxy: false,
                    name: 'Bundle2',
                    isActive: true,
                    isAutomationEnabled: true
                },
            };
            (0, chai_1.expect)(connect_1.getPossibleDebuggerAppKeys.bind(rd)(['*'])).to.eql(['42', '43']);
        });
    });
});
