// appium-remote-debugger/build/test/unit/mixins/message-handlers-specs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const message_handlers_1 = require("../../../lib/mixins/message-handlers");
const helpers_1 = require("../../helpers/helpers");
const remote_debugger_1 = require("../../../lib/remote-debugger");
const chai_1 = require("chai");
describe('connect', function () {
    this.timeout(helpers_1.MOCHA_TIMEOUT);
    let rd;
    this.beforeEach(function () {
        rd = new remote_debugger_1.RemoteDebugger();
    });
    describe('getDebuggerAppKey', function () {
        it('should return the app key for the bundle', function () {
            rd._appDict = {
                ['42']: {
                    id: '42',
                    bundleId: 'io.appium.bundle',
                    isProxy: false,
                    name: 'Bundle',
                    isActive: true,
                    isAutomationEnabled: true
                }
            };
            (0, chai_1.expect)(message_handlers_1.getDebuggerAppKey.bind(rd)('io.appium.bundle')).to.equal('42');
        });
        it('should return the app key for the bundle when proxied', function () {
            rd._appDict = {
                ['42']: {
                    id: '42',
                    bundleId: 'io.appium.bundle',
                    isProxy: false,
                    name: 'Bundle',
                    isActive: true,
                    isAutomationEnabled: true
                },
                ['43']: {
                    id: '43',
                    bundleId: 'io.appium.proxied.bundle',
                    isProxy: true,
                    hostId: '42',
                    name: 'ProxiedBundle',
                    isActive: true,
                    isAutomationEnabled: true
                }
            };
            (0, chai_1.expect)(message_handlers_1.getDebuggerAppKey.bind(rd)('io.appium.bundle')).to.equal('43');
        });
        it('should return undefined when there is no appropriate app', function () {
            rd._appDict = {};
            (0, chai_1.expect)(message_handlers_1.getDebuggerAppKey.bind(rd)('io.appium.bundle')).to.not.exist;
        });
    });
});
